File: models/utilities/finite_difference/verif/SIM_verif/finite_difference.py
import os, sys
import numpy as np

def div_protected(num, denom):
    """
    Division with divide by zero protection
    
    Will replace divide-by-zero conditions with zero
    """
    return np.divide(num, denom, out=np.zeros_like(num), where=(denom != 0))

# Methods here are taken straight from 
# antares/cml/models/utilities/finite_difference/docs/ and are meant to act
# as a comparison to the model's output
class ScalarFirstDerivative(object):
    """
    Calculate estimate of first derivative
    """
    def __init__(self):
        # Initialize values
        self.pass_count = 0
        self.value_im1 = 0
        self.value_im2 = 0
        self.time_im1 = 0
        self.time_im2 = 0

    def compute(self, value, time):
        # Values and times are needed from two time steps ago, so pass
        # until compute is called for at least the 3rd time
        if self.pass_count < 2:
            self.pass_count += 1
            is_good = False
            derivative = 0.0
        else:
            is_good = True
            dt1 = time - self.time_im1
            dt2 = time - self.time_im2
            derivative = div_protected((dt1 + dt2),(dt1*dt2)) * value \
                         + div_protected(dt2,(dt1*(dt1 - dt2))) * self.value_im1 \
                         - div_protected(dt1,(dt2*(dt1 - dt2))) * self.value_im2

        # Assign past values as needed
        # Use float() to avoid any reference issues
        self.value_im2 = float(self.value_im1)
        self.value_im1 = float(value)
        self.time_im2 = float(self.time_im1)
        self.time_im1 = float(time)
        return is_good, derivative

class Vector3FirstDerivative(object):
    """
    Create a list of 3 ScalarFirstDerivative() instantiations
    to represent a 3-vector first derivative
    """
    def __init__(self):
        self.scalar_first_derivatives = [ScalarFirstDerivative(),
                                         ScalarFirstDerivative(),
                                         ScalarFirstDerivative()]

    def compute(self, value, time):
        is_good = True
        derivative = [None]*3
        for i in range(3):
            status, derivative[i] = self.scalar_first_derivatives[i].compute(value[i], time)
            is_good = is_good and status
        return is_good, derivative

class Matrix3x3FirstDerivative(object):
    def __init__(self):
        self.scalar_first_derivatives = [[ScalarFirstDerivative(),
                                         ScalarFirstDerivative(),
                                         ScalarFirstDerivative()],
                                         [ScalarFirstDerivative(),
                                          ScalarFirstDerivative(),
                                          ScalarFirstDerivative()],
                                         [ScalarFirstDerivative(),
                                          ScalarFirstDerivative(),
                                          ScalarFirstDerivative()]]

    def compute(self, value, time):
        is_good = True
        derivative = [[None]*3,[None]*3,[None]*3]
        for i in range(3):
            for j in range(3):
                status, derivative[i][j] = self.scalar_first_derivatives[i][j].compute(value[i][j], time)
                is_good = is_good and status
        return is_good, derivative

class ScalarSecondDerivative(object):
    def __init__(self):
        self.pass_count = 0
        self.value_im1 = 0
        self.value_im2 = 0
        self.value_im3 = 0
        self.time_im1 = 0
        self.time_im2 = 0
        self.time_im3 = 0

    def compute(self, value, time):
        is_good = True

        if self.pass_count < 3:
            self.pass_count += 1
            is_good = False
            derivative = 0.0
        else:
            is_good = True
            dt1 = time - self.time_im1
            dt2 = time - self.time_im2
            dt3 = time - self.time_im3

            derivative = 2*(div_protected((dt2 + dt3),(dt1*dt2*(dt3 - dt1))))*value \
                       - 2*(div_protected((dt1 + dt2),(dt2*dt3*(dt3 - dt1))))*value \
                       + 2*(div_protected((dt2 + dt3),(dt1*(dt1 - dt2)*(dt3 - dt1))))*self.value_im1 \
                       + 2*(div_protected((dt1 + dt3),(dt2*(dt1 - dt2)*(dt2 - dt3))))*self.value_im2 \
                       + 2*(div_protected((dt1 + dt2),(dt3*(dt2 - dt3)*(dt3 - dt1))))*self.value_im3

        self.value_im3 = float(self.value_im2)
        self.time_im3 = float(self.time_im2)
        self.value_im2 = float(self.value_im1)
        self.time_im2 = float(self.time_im1)
        self.value_im1 = float(value)
        self.time_im1 = float(time)

        return is_good, derivative

class Vector3SecondDerivative(object):
    def __init__(self):
        self.scalar_second_derivatives = [ScalarSecondDerivative(),
                                         ScalarSecondDerivative(),
                                         ScalarSecondDerivative()]

    def compute(self, value, time):
        is_good = True
        derivative = [None]*3
        for i in range(3):
            status, derivative[i] = self.scalar_second_derivatives[i].compute(value[i],time)
            is_good = is_good and status
        return is_good, derivative

File: models/utilities/finite_difference/src/finite_difference.cc
/*******************************************************************************
Purpose:
   (Estimate derivatives using finite differences.)

Programmers:
   (((Daniel Matz) (NASA) (July 2017) (Initial version))
    ((Tony Varesic) (OSR) (Apr 2021) (added divide protection)))
*******************************************************************************/

#include "../include/finite_difference.hh"
#include "cml/models/utilities/math_utils/include/math_utils.hh" // for divide_protected()

/******************************************************************************
ScalarFirstDerivative Constructor
******************************************************************************/
ScalarFirstDerivative::ScalarFirstDerivative()
    :
    pass_count(0),
    value_im1(0.0),
    time_im1(0.0),
    value_im2(0.0),
    time_im2(0.0)
{
}

/******************************************************************************
ScalarFirstDerivative::compute
Purpose:Estimate first derivative of a value using backwards difference
******************************************************************************/
bool ScalarFirstDerivative::compute(
    const double value,
    const double time,
    double &derivative)
{
    bool is_good;
    if (pass_count < 2) {
        pass_count++;
        is_good = false;
        derivative = 0.0;
    } else {
        is_good = true;
        const double dt1 = time - time_im1;
        const double dt2 = time - time_im2;
        derivative =   MathUtils::divide_protected(dt1 + dt2 , dt1 * dt2, 0.0, true) * value
                     + MathUtils::divide_protected(dt2 , dt1 * (dt1 - dt2), 0.0, true) * value_im1
                     - MathUtils::divide_protected(dt1 , dt2 * (dt1 - dt2), 0.0, true) * value_im2;

    }
    value_im2 = value_im1;
    time_im2 = time_im1;
    value_im1 = value;
    time_im1 = time;
    return is_good;
}

/******************************************************************************
Vector3FirstDerivative::compute
Purpose:Estimate first derivative of a vector using backwards difference
******************************************************************************/
bool Vector3FirstDerivative::compute(
    const double (&value)[3],
    const double time,
    double (&derivative)[3])
{
    bool is_good = true;
    for (int i = 0; i < 3; i++) {
        bool status = scalar_first_derivative[i].compute(value[i], time, derivative[i]);
        is_good = is_good and status;
    }
    return is_good;
}

/******************************************************************************
Matrix3x3FirstDerivative::compute
Purpose:Estimate first derivative of a matrix using backwards difference
******************************************************************************/
bool Matrix3x3FirstDerivative::compute(
    const double (&value)[3][3],
    const double time,
    double (&derivative)[3][3])
{
    bool is_good = true;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            bool status = scalar_first_derivative[i][j].compute(value[i][j], time, derivative[i][j]);
            is_good = is_good and status;
        }
    }
    return is_good;
}

/******************************************************************************
ScalarSecondDerivative Constructor
******************************************************************************/
ScalarSecondDerivative::ScalarSecondDerivative()
    :
    pass_count(0),
    value_im1(0.0),
    time_im1(0.0),
    value_im2(0.0),
    time_im2(0.0),
    value_im3(0.0),
    time_im3(0.0)
{
}

/******************************************************************************
ScalarSecondDerivative::compute
Purpose:Estimate second derivative of a value using backwards difference
******************************************************************************/
bool ScalarSecondDerivative::compute(
    const double value,
    const double time,
    double &derivative)
{
    bool is_good;
    if (pass_count < 3) {
        pass_count++;
        is_good = false;
        derivative = 0.0;
    } else {
        is_good = true;
        const double dt1 = time - time_im1;
        const double dt2 = time - time_im2;
        const double dt3 = time - time_im3;
        derivative =   2 * MathUtils::divide_protected(dt2 + dt3 , dt1 * dt2 * (dt3 - dt1), 0.0, true) * value
                     - 2 * MathUtils::divide_protected(dt1 + dt2 , dt2 * dt3 * (dt3 - dt1), 0.0, true) * value
                     + 2 * MathUtils::divide_protected(dt2 + dt3 , dt1 * (dt1 - dt2) * (dt3 - dt1), 0.0, true) * value_im1
                     + 2 * MathUtils::divide_protected(dt1 + dt3 , dt2 * (dt1 - dt2) * (dt2 - dt3), 0.0, true) * value_im2
                     + 2 * MathUtils::divide_protected(dt1 + dt2 , dt3 * (dt2 - dt3) * (dt3 - dt1), 0.0, true) * value_im3;
    }
    value_im3 = value_im2;
    time_im3 = time_im2;
    value_im2 = value_im1;
    time_im2 = time_im1;
    value_im1 = value;
    time_im1 = time;
    return is_good;
}

/******************************************************************************
Vector3SecondDerivative::compute
Purpose:Estimate second derivative of a vector using backwards difference
******************************************************************************/
bool Vector3SecondDerivative::compute(
    const double (&value)[3],
    const double time,
    double (&derivative)[3])
{
    bool is_good = true;
    for (int i = 0; i < 3; i++) {
        bool status = scalar_second_derivative[i].compute(value[i], time, derivative[i]);
        is_good = is_good and status;
    }

    return is_good;
}

